### ISSUE TITLE
How to get the current location from Svelvet to use as initialLocation

### ISSUE THREAD

**Issue by dilignt** [open-source-labs/Svelvet#186]
I reactively rerender svelvet when my list of nodes and edges are modified by user interaction of widgets external to Svelvet.  This works well, although I'm struggling with the initialLocation prop on the Svelvet component.

When the user pans and zooms their initialLocation and scale needs to change to reflect their zooming so that when the Svelvet component rerenders it doesn't jump back to the start postion in the middle of the screen.  I modified the d3Scale store to pass back the entire d3 zoom transform instead of just the scale part of the transform.  The scale is easy to modify to the correct initialScale number value, but I can't figure out how to change the rest of the transform to get the correct x and y values.  

Hopefully this code illustrates what I am trying to do

```
import { coreSvelvetStore } from "svelvet/stores/store";

const store = coreSvelvetStore.d3Scale;   // d3Scale modified to store the entire zoom transform, not just the k part of it, so this code won't work out-of-the-box.
let startPos = { x: window.innerWidth / 2, y: window.innerHeight / 2 };
let scale = 2;

store.subcribe( value => setTransform(value) );     // Zoom transform {x,y,k} passed via store from handleZoom function

function setTransform(t) {
    scale = {t.k - 0.4)/0.16;     // reverse the initZoom function scale calculation

    // What goes here to modify startPos x and y using t.x and t.y from the handleZoom transform?
}

<Svelvet initialLocation={startPos} initialScale={scale} />     // other required props omitted

```

I'm not good enough with d3 to figure out what to do with the passed transform to maintain the startPos at the centre of the screen after the user has panned and zoomed.

It would be good if there was 2-way binding supported for the initialLocation and initialZoom props (i.e. they are not just initial values) so that they could update automatically as the user pans and zooms, but in the abscence of this I'm trying to get this to work.

Hope this makes sense, any help greatly appreciated
many thanks
Mike

**Comment by michael-chiang-dev5** [comment]
Hi @dilignt

Thank you for your input! We have implemented some developer functionality to make interacting with d3 parameters easier. You can use functions getD3PositionX, getD3PositionY, getD3Zoom to get d3 parameters. Note that zoom is based on reference point (0,0) and might explain some of challenges working with d3 parameters directly in thestore.

Test it here: https://stackblitz.com/edit/data-reactivity?file=src/routes/+layout.svelte

Closing ticket, but if you have any questions / suggestions please don't hesitate to reopen or open new ticket. Thank you for supporting Svelvet!

**Comment by dilignt** [comment]
Hi @michael-chiang-dev5 

Thanks for your efforts in trying to provide a solution for me.  I've tried to try it out but I ran into problems upgrading to version 6.0.4 - I get this exception thrown when I'm trying to upgrade my diagram.  I've got hundreds of nodes in my app and I've built a layer of interactivity above svelvet to add and remove them dynamically, so I'm not in a position to start again from scratch.  From looking at the code within the exception stack I tried adding a width and a height to my node css class but that hasn't worked :(

```
Node.js:67 Uncaught TypeError: Cannot read properties of undefined (reading 'setPositionFromMovement')
    at Node.js:67:40
    at Object.update (index.mjs:42:13)
    at Node.setPositionFromMovement (Node.js:64:20)
    at Node.js:67:40
    at Object.update (index.mjs:42:13)
    at Node.setPositionFromMovement (Node.js:64:20)
    at forceCssHeightAndWidth (getCss.js:45:10)
    at Node.svelte:21:9
    at flush (index.mjs:1175:17)
    at init (index.mjs:2034:9)
```
The node diagram draws in before the exception and I notice that there are now unwanted mouseover effects on the edges and also the handles are back - I removed them in v5.0.0 by doing this but it no longer works and I now can't find the elements in the dom browser to search for :(
```
	:global(.Anchor-inert) {
		visibility: hidden !important;
	}
```
For the moment I need to revert to v5.0.0 because my app is now going into beta and I need it to be working, but hopefully I'll be able to upgrade to a newer version of Svelvet when the upgrade path is smoother.

I worked around the location and zoom issue by adding start and end handlers to the d3 zoom function that passed the entire transform to my client svelte app, and then calculated reversed deltas of x and y between start and end transforms depending on whether they are mouse events or wheel events to maintain the initialLocation.  It pretty much works for pan although not exactly on zoom wheel events, and the coordinate calculations are different depending on pan or zoom, but I've got it quite close to be usable.  I presume you do this in v6.0.4 to maintain the values for your new functions?

Thanks again, I appreciate the time you've taken looking to resolve this issue.  Although I think it would be much easier for the developer to have two-way binding on initialLocation and initialZoom props that automatically instead of needing to perform calculations wrt 0,0 coordinates on function calls.

cheers
